### ISSUE TITLE
Assertion failed with `new-build` and `--allow-newer`

### ISSUE THREAD

**Issue by tom-bop** [haskell/cabal#4636]
With GHC 8.2.1:

```
$ cabal --version
cabal-install version 1.24.0.2
compiled using version 1.24.2.0 of the Cabal library
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.2.1
$ cabal new-build -j broken.cabal --allow-newer=base,time
Assertion failed
CallStack (from HasCallStack):
assert, called at ./Distribution/Client/InstallPlan.hs:508:31 in
main:Distribution.Client.InstallPlan
```

Note `--allow-newer=base` works fine.

Here's the .cabal file:

```
name: broken
version: 0.1
build-type:          Simple
cabal-version:       >=1.10

library
  exposed-modules:
  build-depends:
      base
  hs-source-dirs:      src
  default-language:    Haskell2010
```

(I've simplified this to its minimum repro, but it also fails when it really has a transitive dependency on `time`)

**Comment by hvr** [comment]
I tried locally but couldn't reproduce this. Does this actually reproduce with cabal-install-2.0 or -2.1?

**Comment by tom-bop** [comment]
Unfortunately I can only confirm with 1.24.0.2 at the moment. I can try to test with 2.0 in the next couple of days.

**Comment by hvr** [comment]
I wonder if this is related or was even fixed already by @ezyang's af24cefe12adf0f0e1745b70db45507ae5b45d06

**Comment by robrix** [comment]
My team is running into this (likewise with cabal-install 1.24.0.2), but only on our (x86-64 Linux) CI/build machines, not our (x86-64 Mac OS X) local development machines. It likewise seems to have started with the addition of a dependency on `time`.

I’ll attempt a repro with a newer `cabal-install`.

**Comment by robrix** [comment]
Can confirm that this does not happen at least as of 41a4ef1cb7d52de04e49ca51a4815ac247eaabc4 (the current HEAD of the 2.0 branch) 👍

**Comment by 23Skidoo** [comment]
Closing as fixed by af24cefe12adf0f0e1745b70db45507ae5b45d06 then.
